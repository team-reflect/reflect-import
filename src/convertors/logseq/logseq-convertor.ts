import {DOM, domArrayToHtml, domToHtml} from 'helpers/dom'
import {header1, list, listItem} from 'helpers/generators'
import {markdownToHtml} from 'helpers/markdown'
import {validateNotes} from 'helpers/validate'

import {tryParseTime, toLogseqId, assertsLogseqNoteBlock} from './logseq-helpers'
import {logseqPropertiesToMarkdown} from './logseq-properties'
import {exportSchema} from './schema'
import {LogseqBlock, LogseqConversionError, LogseqExport, LogseqNote} from './types'
import {Convertor} from '../../convertor'
import {Backlink, ConvertedNote, ConvertOptions, ConvertResponse} from '../../types'

export class LogseqConvertor extends Convertor {
  static accept = {'application/json': ['.json']}
  static description = 'Logseq JSON'

  noteIds: Record<string, string> = {}

  async convert({data}: ConvertOptions): Promise<ConvertResponse> {
    const parsed: LogseqExport = JSON.parse(data)
    const validated = await exportSchema.parseAsync(parsed)

    // Create a map of page names to ids.  We use this for the link resolver and adding
    // ids to backlinks.
    this.noteIds = validated.blocks.reduce(
      (acc, note) => ({
        ...acc,
        [note['page-name']]: toLogseqId(note.id, note['page-name']),
      }),
      {},
    )

    // Convert all notes except for whiteboard pages
    const convertedNotes = validated.blocks.flatMap((note) => {
      if (note?.properties?.['ls-type'] === 'whiteboard-page') {
        return []
      }
      return [this.convertLogseqNote(note)]
    })

    return validateNotes(convertedNotes)
  }

  /**
   * Converts an individual logseq note into a ConvertedNote.  A logseq note has
   * the page name and the blocks (children).  We convert all the children to HTML
   * and then add the page name as an H1 to the top of the note.
   */
  private convertLogseqNote(note: LogseqNote): ConvertedNote {
    const {id, children} = note
    const subject = note['page-name']

    // Get all the HTML and backlinks for the current note
    const {html, backlinks} = this.parseBlocks(children)
    // Update the backlinks with the note ids from the page name to id map
    const updatedBacklinks = backlinks.map((b) => ({
      label: b.label,
      id: toLogseqId(this.noteIds[b.label], b.label),
    }))

    return {
      id: toLogseqId(id, subject),
      html: domArrayToHtml([header1(subject), html]),
      subject,
      backlinks: updatedBacklinks,
      dailyAt: tryParseTime(subject),
    }
  }

  /**
   * Converts a list of logseq blocks into HTML.  This is a recursive through the
   * makeHtml function.  Each block is an li in a ul with the block's content in the
   * li.
   */
  private parseBlocks(blocks: LogseqBlock[]): {html: string; backlinks: Backlink[]} {
    const blockItems: DOM[] = []
    const backlinkCollection: Backlink[] = []

    // Get the HTML and backlinks for each block
    for (const block of blocks) {
      const {html, backlinks} = this.makeHtml(block)
      blockItems.push(listItem(html))
      backlinkCollection.push(...backlinks)
    }

    return {
      html: domToHtml(list(blockItems)),
      backlinks: backlinkCollection,
    }
  }

  /**
   * Converts a single logseq block's content into HTML.  This will recursively
   * call parseBlocks if the block has children.
   */
  private makeHtml(block: LogseqBlock): {html: string; backlinks: Backlink[]} {
    assertsLogseqNoteBlock(block)
    // Logseq supports markdown and org.  I think doing the org conversion will be
    // more difficult.  So I'm going to start with markdown.
    if (block.format !== 'markdown') {
      throw new LogseqConversionError(
        'Logseq importer is only able to import markdown blocks',
      )
    }

    // If the block has content then we use that.
    let blockContent = block.content
    // If a block does not have content then it might have properties and we'll
    // convert those properties to markdown.  In the case of frontmatter it will
    // have both.  But we will ignore the properties and just use the frontmatter
    // as markdown.
    if (!blockContent && block.properties) {
      blockContent = logseqPropertiesToMarkdown(block.properties, this.noteIds)
    }

    // Get the data for the current block
    let {html, backlinks} = markdownToHtml(blockContent || '', {
      graphId: this.graphId,
      linkHost: this.linkHost,
      pageResolver: (pageName) => toLogseqId(this.noteIds[pageName], pageName),
    })

    // If the block has children then we need to get the html for the children
    if (block.children.length) {
      const {html: childHtml, backlinks: childBacklinks} = this.parseBlocks(
        block.children,
      )
      // We just append the child html to this current block HTML
      html = domArrayToHtml([html, childHtml])
      backlinks = [...backlinks, ...childBacklinks]
    }

    return {
      html,
      backlinks,
    }
  }
}
