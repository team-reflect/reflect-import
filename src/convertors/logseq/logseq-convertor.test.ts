import {describe, it, expect} from 'vitest'

import exampleExport from './fixtures/logseq-example.json'
import exampleOrgExport from './fixtures/logseq-org-example.json'
import examplePropertiesExport from './fixtures/logseq-property-example.json'
import exampleTodoExport from './fixtures/logseq-todo-example.json'
import exampleWhiteboardExport from './fixtures/logseq-whiteboard-example.json'
import {LogseqConvertor} from './logseq-convertor'

describe('LogseqConvertor', () => {
  it('parses example', async () => {
    const convertor = new LogseqConvertor({graphId: '123'})
    const {notes} = await convertor.convert({data: JSON.stringify(exampleExport)})

    expect(notes).toMatchSnapshot()
  })
  it('parses properties example', async () => {
    const convertor = new LogseqConvertor({graphId: '123'})
    const {notes} = await convertor.convert({
      data: JSON.stringify(examplePropertiesExport),
    })

    expect(notes).toMatchSnapshot()
  })
  it('parses TODO example', async () => {
    const convertor = new LogseqConvertor({graphId: '123'})
    const {notes} = await convertor.convert({data: JSON.stringify(exampleTodoExport)})

    expect(notes).toMatchSnapshot()
  })
  it('handles and ignores whiteboard pages', async () => {
    const convertor = new LogseqConvertor({graphId: '123'})
    const {notes} = await convertor.convert({
      data: JSON.stringify(exampleWhiteboardExport),
    })

    expect(notes).toMatchSnapshot()
  })
  it('exception for Org example', () => {
    const convertor = new LogseqConvertor({graphId: '123'})
    expect(
      async () => await convertor.convert({data: JSON.stringify(exampleOrgExport)}),
    ).rejects.toThrowError()
  })
})
