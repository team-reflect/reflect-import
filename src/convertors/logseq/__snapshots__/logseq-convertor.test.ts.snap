// Vitest Snapshot v1

exports[`LogseqConvertor > handles and ignores whiteboard pages 1`] = `
[
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Contents</h1><ul><li></li></ul>",
    "id": "logseq-f543cf64-48b3-440e-816e-cbdeff1db261",
    "subject": "Contents",
  },
  {
    "backlinks": [],
    "dailyAt": 1685491200000,
    "html": "<h1>May 31st, 2023</h1><ul><li><p>This is a test with a whiteboard.</p></li></ul>",
    "id": "31052023",
    "subject": "May 31st, 2023",
  },
]
`;

exports[`LogseqConvertor > parses TODO example 1`] = `
[
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Contents</h1><ul><li></li></ul>",
    "id": "logseq-6456a448-562e-4481-b7d9-3609f5271d5e",
    "subject": "Contents",
  },
  {
    "backlinks": [],
    "dailyAt": 1683331200000,
    "html": "<h1>May 6th, 2023</h1><ul><li><p>TODO Thing 1
:LOGBOOK:
CLOCK: [2023-05-06 Sat 14:02:38]--[2023-05-06 Sat 14:02:40] =>  00:00:02
:END:</p></li><li><p>DOING Thing 2
:LOGBOOK:
CLOCK: [2023-05-06 Sat 14:03:11]
:END:</p></li><li><p>DONE Thing 3
:LOGBOOK:
CLOCK: [2023-05-06 Sat 14:03:14]--[2023-05-06 Sat 14:03:15] =>  00:00:01
CLOCK: [2023-05-06 Sat 14:03:16]--[2023-05-06 Sat 14:03:19] =>  00:00:03
CLOCK: [2023-05-06 Sat 14:03:22]--[2023-05-06 Sat 14:03:26] =>  00:00:04
CLOCK: [2023-05-06 Sat 14:03:28]--[2023-05-06 Sat 14:03:29] =>  00:00:01
:END:</p></li><li></li><li></li><li><p>TODO</p></li><li></li></ul>",
    "id": "06052023",
    "subject": "May 6th, 2023",
  },
]
`;

exports[`LogseqConvertor > parses example 1`] = `
[
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Contents</h1><ul><li></li></ul>",
    "id": "logseq-048b84ca-e374-40c9-90ba-1129207b1d88",
    "subject": "Contents",
  },
  {
    "backlinks": [
      {
        "id": "02052023",
        "label": "May 2nd, 2023",
      },
      {
        "id": "04052023",
        "label": "May 4th, 2023",
      },
    ],
    "dailyAt": 1683072000000,
    "html": "<h1>May 3rd, 2023</h1><ul><li><p>This is a test graph</p></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/02052023\\">May 2nd, 2023</a></p><ul><li><p>This is some notes for yesterday</p></li></ul></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/04052023\\">May 4th, 2023</a></p><ul><li><p>This is some notes for tomorrow</p></li></ul></li></ul>",
    "id": "03052023",
    "subject": "May 3rd, 2023",
  },
  {
    "backlinks": [],
    "dailyAt": 1682985600000,
    "html": "<h1>May 2nd, 2023</h1><ul><li><p>This is yesterday</p></li></ul>",
    "id": "02052023",
    "subject": "May 2nd, 2023",
  },
  {
    "backlinks": [
      {
        "id": "logseq-645304ab-3a8e-41f6-996c-3e023a731b63",
        "label": "Another Page",
      },
    ],
    "dailyAt": undefined,
    "html": "<h1>New Page</h1><ul><li><p>This is a standalone page.</p></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/logseq-645304ab-3a8e-41f6-996c-3e023a731b63\\">Another Page</a></p></li><li><p>:properties:</p></li><li></li></ul>",
    "id": "logseq-6453049e-5817-4a48-8ad2-8a64c56295d6",
    "subject": "New Page",
  },
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Another Page</h1><ul><li><ul>
<li>property: page</li>
<li>thing: yes</li>
</ul></li><li><p>Deeply</p><ul><li><p>Nested</p><ul><li><p>Text</p><ul><li><p>Is</p><ul><li><p>Here</p></li></ul></li></ul></li></ul></li></ul></li><li><p><a href=\\"http://example.com\\">external link</a></p></li><li><p><img src=\\"../assets/SuccessKid_1683162562594_0.jpg\\" alt=\\"SuccessKid.jpg\\"></p></li><li><p>By Originally from the mother's Flickr page, found on Know Your Meme, Fair use, <a href=\\"https://en.wikipedia.org/w/index.php?curid=46523616\\">https://en.wikipedia.org/w/index.php?curid=46523616</a></p></li></ul>",
    "id": "logseq-645304ab-3a8e-41f6-996c-3e023a731b63",
    "subject": "Another Page",
  },
]
`;

exports[`LogseqConvertor > parses properties example 1`] = `
[
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Contents</h1><ul><li></li></ul>",
    "id": "logseq-048b84ca-e374-40c9-90ba-1129207b1d88",
    "subject": "Contents",
  },
  {
    "backlinks": [
      {
        "id": "logseq-6455bc66-8886-449b-bddc-12c53f30b1a5",
        "label": "Company1",
      },
    ],
    "dailyAt": undefined,
    "html": "<h1>Person1</h1><ul><li><ul>
<li>company: <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/logseq-6455bc66-8886-449b-bddc-12c53f30b1a5\\">Company1</a></li>
<li>phone: 12345</li>
<li>jobtitle: manager</li>
</ul></li><li><p>This is person 1</p></li><li></li><li></li></ul>",
    "id": "logseq-6455adda-ef16-45a3-9cb3-bd8a5f6dd831",
    "subject": "Person1",
  },
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Company1</h1><ul><li><p>This is a page for company one</p></li></ul>",
    "id": "logseq-6455bc66-8886-449b-bddc-12c53f30b1a5",
    "subject": "Company1",
  },
  {
    "backlinks": [
      {
        "id": "logseq-6455aef4-7231-4932-8679-cce8b3403363",
        "label": "Company Two",
      },
    ],
    "dailyAt": undefined,
    "html": "<h1>Person2</h1><ul><li><ul>
<li>company: <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/logseq-6455aef4-7231-4932-8679-cce8b3403363\\">Company Two</a></li>
<li>phone: 12345</li>
<li>jobtitle: manager</li>
</ul></li><li></li><li><p>This is person 2</p></li></ul>",
    "id": "logseq-6455ae48-1216-4b6c-bb26-d3e67adde877",
    "subject": "Person2",
  },
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Company Two</h1><ul><li><p><a href=\\"https://reflect.app/g/123/tag/company\\">#company</a></p></li><li></li></ul>",
    "id": "logseq-6455aef4-7231-4932-8679-cce8b3403363",
    "subject": "Company Two",
  },
  {
    "backlinks": [
      {
        "id": "logseq-6455aef4-7231-4932-8679-cce8b3403363",
        "label": "Company Two",
      },
      {
        "id": "logseq-6455bc66-8886-449b-bddc-12c53f30b1a5",
        "label": "Company1",
      },
    ],
    "dailyAt": undefined,
    "html": "<h1>Person3</h1><ul><li><ul>
<li>company: <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/logseq-6455aef4-7231-4932-8679-cce8b3403363\\">Company Two</a>, <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/logseq-6455bc66-8886-449b-bddc-12c53f30b1a5\\">Company1</a></li>
</ul></li><li><p>This person is associated with both companies</p></li></ul>",
    "id": "logseq-6455b0ff-0bf2-4965-b579-ae86af4801ab",
    "subject": "Person3",
  },
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Some Page</h1><ul><li><hr>
<h2>title: sample</h2></li><li></li><li><p>Frontmater properties</p></li></ul>",
    "id": "logseq-6455b17c-958f-4590-bc17-fe48676ecf23",
    "subject": "Some Page",
  },
]
`;
